### ISSUE TITLE
Migrate to repo to "src" layout

### ISSUE THREAD

**Issue by bryevdv** [bokeh/bokeh#11507]
cc @bokeh/dev

We have had (and seem to continue to have) issues with testing real package installs in CI due to using the traditional `bokeh/bokeh` directory layout for the repo. This has also caused issues with test coverage measurement in the past. Current best practice recommendations seem to increasingly move away from the old recommended `project/project` structure, in favor of a `project/src` structure [1], [2] Chiefly, doing this helps avoid the possibility of accidentally running code from the repo. 

Since this is a non-trivial change that will require care and testing and communication, I propose to make the change for the 3.0 release cycle. 

[1] https://blog.ionelmc.ro/2014/05/25/python-packaging/#the-structure
[2] https://hynek.me/articles/testing-packaging/

**Comment by ianthomas23** [comment]
+1 on this.  I ended up taking a similar approach with `contourpy` (but using `lib` directory as `src` is where I've put the C/C++ source code)  and it works well for editable installs.

Would `bokehjs` move to `src` or stay at the top level?

**Comment by mattpap** [comment]
> Would bokehjs move to src or stay at the top level?

bokehjs has its own layout (src equivalent), so most likely yes.

**Comment by bryevdv** [comment]
If we wanted to go whole hog, the "best" layout is probably to split by langauge at the top level:
```
+ README.md
+ javascript
|   + package.json
|   + src
|       + lib
+ python
    + setup.py
    + src
        + __init__.py
```
I don't have strong feelings about this, but I would definitely support this move if others also agree.

**Comment by bryevdv** [comment]
I've thought about this some more. I do personally very much favor the structure above (split by language at the top level). I think it is the cleanest and most "correct" layout, without question. However, I am also pragmatic, and I think that breaking ~10 years of accumulated links to examples, etc. out in the wild (that can't be redirected) is bad for users in a way that cannot be justified merely by our preferences. 

So instead I think we should settle for this more more modest change:

```
+ README.md
+ bokehjs
|   + node_modules
|   + package.json
|   + src
|   |  + lib
|   + tests
+ examples
+ setup.py
+ src
|   + __init__.py
+ tests
+ scripts
+ sphinx
```

TLDR; limit to just only `bokeh/bokeh` → `bokeh/src` 

The main question is when. Even this more minimal change will necessitate a good degree of quiescence in the repo for at least a few days, and possibly closing stale PRs, etc. @mattpap you will need to communicate what will work for you.

**Comment by mattpap** [comment]
> TLDR; limit to just only `bokeh/bokeh` → `bokeh/src`

Seems fine me. I don't see this being very conflicting with existing PRs.

**Comment by bryevdv** [comment]
Well, maybe the full language split is workable if we factor examples separately? @mattpap what do you think about
```
+ README.md
+ examples
|   + bokehjs
|   + plotting
+ javascript
|   + node_modules
|   + package.json
|   + src
|   |  + lib
|   + tests
+ python
|   + setup.py
|   + src
|   |   + __init__.py
|   + tests
+ scripts
+ sphinx
```

**Comment by mattpap** [comment]
Are there any technical reasons this:

> + javascript

to be beneficial? If not, I don't see a point.

Also, how this:

> + examples
> |   + bokehjs

is supposed to work? More likely than not, this will require tooling from bokehjs, which I don't intend to replicate outside bokehjs. Actually, pure bokehjs' end-user examples (not what we have e.g. in integration tests) would be best kept in a separate [repository](https://github.com/bokeh/bokehjs-examples), because we should be pointing to bokehjs' package in examples' `package.json` (not the build directory), to make them self-contained. Having them in the source repository, would make things inconsistent, because we can't provide bokehjs packages for each pull request (we even still don't have nightly builds).
